README.md

```diff
@@ -1,18 +1,129 @@
+
+* how to generate the effect.
+* how the effect is done
+
+For the basic effects, the types of those functions clearly state the intent of the effect.
+
+In this section, the utility functions of the effects are not listed. Once you understand the basic usage of an effect, the type of any utility function typically explains what is does.
+
+Of course, the best learning progress is done by experimenting inside `ghci`.
+
+### The Exception Effect
+
+The exception effect adds the possibility to exit a computation preemptively with an exception. Note that the exceptions from this library are handled by the programmer and have nothing to do with exceptions thrown inside the Haskell run-time.
+
+```haskell
+throwError :: Member (Exc e) r => e -> Eff r a
+runError :: Eff (Exc e ': r) a -> Eff r (Either e a)
+```
+
+An exception can be thrown using the `throwError` function. Its return type is `Eff r a` with an arbitrary type `a`. When handling the effect, the result-type changes to `Either e a` instead of
+just `a`. This indicates how the effect is handled: The returned value is either the thrown exception or the value returned.
+
+### The State Effect
+
+The state effect provides readable and writable state during a computation.
+
+```haskell
+get :: Member (State s) r => Eff r s
+put :: Member (State s) r => s -> Eff r ()
+modify :: Member (State s) r => (s -> s) -> Eff r ()
+runState :: s -> Eff (State s ': r) a -> Eff r (a, s)
+```
 
-  * Effects can be added, removed, and interwoven without changes to code not
-    dealing with those effects.
+There are different functions for programming with the state-effect. The `get` functions accesses the current state and makes it usable within the further computation. The `put` function sets the state to the given value. `modify` updates the state using a mapping function.
+
+The state-effect is handled using `runState` function. It takes the initial state as argument and returns the final state and effect-result.
+
+### The Reader Effect
+
+The reader effect provides an environment that can be read. Sometimes it is considered as read-only state that stays same during the computation.
+
+```haskell
+ask :: Member (Reader e) r => e -> Eff r e
+runReader :: e -> Eff (Reader e ': r) a -> Eff r a
+```
+
+The environment given to the handle the reader effect is the one given during the computation if asked for.
+
+There are two variants of readers: strict and lazy. Each has its own module and provide the same interface. By importing one or the other, it can be controlled if the reader is strict or lazy in its environment argument.
+
+### The Writer Effect
+
+The writer effect allows to output messages during a computation. It is sometimes referred to as write-only state, which gets retrieved.
+
+```haskell
+tell :: Member (Writer w) r => w -> Eff r ()
+runWriter :: (w -> b -> b) -> b -> Eff (Writer w ': r) a -> Eff r (a, b)
+runListWriter :: Eff (Writer w ': r) a -> Eff r (a, [w])
```

Comment:
@power-fungus using above template for describing effects:
- `tell`: single type of request to handler. corresponds to `Tell` case of associated GADT
- `runWriter` and `runListWriter` being two distinct handlers/interpreters (we have the choice of picking one when we want to run the computation (i.e., call-site)